client/src/modules/purchases/purchases.service.js

```diff
@@ -157,5 +158,37 @@ function PurchaseOrderService(
     return $httpParamSerializer(options);
   }
 
+  /**
+   * @method formatOptimalPurchase
+   *
+   * This Service analyzes for all existing repositories, Inventories that have reached a point of order,
+   * then calculates the overall quantity ordered for each Inventory
+   *
+   * @param {*} inventories - This is the list of all inventories
+   * @param {*} stock - it is the data of the inventories having reached their point of order
+   */
+  function formatOptimalPurchase(inventories, stock) {
+    inventories.forEach(inventory => {
```

Comment:
This looks a lot more complicated than it needs to be - a lot of these properties and calculations are done already by the [`PurchaseOrderForm`](https://github.com/IMA-WorldHealth/bhima/blob/master/client/src/modules/purchases/create/PurchaseOrderForm.js) and the [`PurchaseOrderItemService`](https://github.com/IMA-WorldHealth/bhima/blob/master/client/src/modules/purchases/create/PurchaseOrderItemService.js.

Could you try using those?  I would proceed like this:

   1. Start by moving this function into the `PurchaseOrderFormService`, since it will only be used on the create page.
   2. Use the `PurchaseOrderItemService` for each item.  That would look like this, I think:
```js
inventories.forEach(inventory => {
  const row = new PurchaseOrderItem(inventory);
  // quantity is by default 1, so set it it 0
  row.quantity = 0;

  stock.forEach(item => {
    if (row.inventory_uuid === item.inventory_uuid) {
      row.quantity += item.S_Q;
    }
  });
```
   3. Then finally, add the custom logic to sort:
```js
     const optimized = rows.filter(inventory => inventory.quantity > 0); 
     const optimizedSorted = optimized.sort((a, b) => Number(b.quantity) - Number(a.quantity));
     return optimizedSorted;
```

Reply:
If you do all this, you won't need to do a `forEach` loop in the `optimalPurchase()` function.  You can just push these into the stack.

```js
const optimalPurchaseData = Purchases.formatOptimalPurchase(vm.inventories, rows); 

// clear the grid as suggested above
vm.order.clear();

optimalPurchaseData.forEach(item => {
  vm.order.store.post(item);
});

vm.order.digest();
```

Note that if you make this a function of `PurchaserOrderForm`, instead of `vm.order`, you can just use `this`.